Detect separate git dir and set git config path value appropriately

cmrd-senya opened the pull request:
##### SUMMARY

fixes #30034 

Detects a separate git tree configuration in the working tree  and sets `$GIT_DIR` environment variable to that value

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
git.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.0.0.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

Commit:
Detect separate git dir and set $GIT_DIR appropriately

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -1012,13 +1012,22 @@ def main():
     # call run_command()
     module.run_command_environ_update = dict(LANG='C', LC_ALL='C', LC_MESSAGES='C', LC_CTYPE='C')
 
+    separate_git_dir = None
```
There's no point of having this var here unless you're going to reuse it after the checks. Please remove this line.

cmrd-senya commented:
I updated the PR. There is a failing test: https://app.shippable.com/github/ansible/ansible/runs/59445/27/tests

Looks like it's not my fault:
```
fatal: unable to access 'https://github.com/jimi-c/test_role/': Failed to connect to github.com port 443: Operation timed out
```

mattclay commented:
I've restarted CI for this PR.

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -1019,6 +1019,14 @@ def main():
+        elif os.path.isfile(os.path.join(dest, '.git')):
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+                match = re.match('gitdir: (.+)$', data)
```
Please dedent all the following code, so that it would be outside of `with`-block.

cmrd-senya commented:
@webknjaz, updated

webknjaz commented:
@cmrd-senya could you please add tests for this?
https://github.com/ansible/ansible/tree/devel/test/integration/targets/git/tasks

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -1019,6 +1019,14 @@ def main():
+        elif os.path.isfile(os.path.join(dest, '.git')):
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+            match = re.match('gitdir: (.+)$', data)
```
should the line start with `gitdir` or should it be in the end of line with smth prefixed? you might want to prefix the regex with `^`.

On the other hand, taking into account that regexps are expensive, you could just use str.split or partition with bits of validation on top, like:
```python
try:
    ref_prefix, separate_git_dir = data.split('gitdir: ', 1)
    if ref_prefix:
        raise ValueError('.git file has invalid git dir reference format')
    if not os.path.isdir(separate_git_dir):
        raise TypeError('%s is not a directory' % separate_git_dir)
    os.environ['GIT_DIR'] = separate_git_dir
except (TypeError, ValueError) as err:
    """``.git`` file does not have a valid format for detached Git dir."""
    module.fail_json(
        msg='Current repo does not have a valid reference to a '
        'separate Git dir or it refers to the invalid path',
        details=str(err),
    )
```

cmrd-senya replied:
Do I have to write test cases for validations then?

cmrd-senya replied:
Also, what is the better way to handle the exceptions: to keep trying the execution or to shut down immediately?

webknjaz replied:
Fail fast and loud for fatal errors like being unable to fulfill the module's purpose. I've updated a snippet in the comment above to reflect this.
And yes, negative test cases for error code paths are highly appreciated, please add them.

webknjaz replied:
@cmrd-senya I've updated your branch with my implementation for now.

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -1019,6 +1019,14 @@ def main():
+            match = re.match('gitdir: (.+)$', data)
+            if match:
+                separate_git_dir = match.group(1)
+                os.environ['GIT_DIR'] = separate_git_dir
```
I would not assign it w/o validation:
1) it might cause errors happening in some other place, which would be hard to spot
2) it might be a weak place, vulnerable to injection attacks (security issue)

webknjaz replied:
@cmrd-senya I've got concerns about modifying globally accessible env vars, so after talking to a colleague he brought up that whenever you change the env var value you should revert it to its initial state after using it.
So I strongly recommend moving to CLI arg instead of messing with env vars.

webknjaz commented on test/integration/targets/git/tasks/main.yml:
```diff
@@ -36,3 +36,8 @@
 - include_tasks: archive.yml
+- include_tasks: separate-git-dir.yml
+  when:
+    - not gpg_version.stderr
```
I don't think it has anything to do with GPG commit signing.

webknjaz commented on test/integration/targets/git/tasks/separate-git-dir.yml:
```diff
@@ -0,0 +1,18 @@
+    path: '{{ checkout_dir }}'
+
+- name: SEPARATE-GIT-DIR | clone with a separate git dir
+  command: git clone {{ repo_format1 }} git --separate-git-dir=separate.git
```
For safety reasons, I'd create some random dir using `mktemp -d`

cmrd-senya replied:
A random dir for `separate.git` directory? What kind of safety do you mean? If the problem is that it may already exist then we can clear the directory the same way as the `checkout_dir`.

webknjaz replied:
yes, i'm concerned about it already existing in place, but also current approach will cause conflicts and false failures in case of simultaneous test runs.

webknjaz replied:
Why not use `{{ checkout_dir }}` instead of `git` arg + chdir?

webknjaz replied:
After adding support for module arg, you'll need to just use the git module itself rather then shell command.

webknjaz commented on test/integration/targets/git/tasks/separate-git-dir.yml:
```diff
@@ -0,0 +1,18 @@
+- name: SEPARATE-GIT-DIR | update repo the usual way
+  git:
+    repo: "{{ repo_format1 }}"
+    dest: "{{ checkout_dir }}"
```
Also don't forget to clean-up dirs after testing (tear down stage)

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -1019,6 +1019,14 @@ def main():
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+            match = re.match('gitdir: (.+)$', data)
+            if match:
```
Also, in case of not match `gitconfig` would have `None` value, which would've broken this module in some place being executed later.

webknjaz committed:
Rewrite gitdir processing snippet with validation

webknjaz committed:
Restore gitconfig path assignment

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -1019,6 +1019,23 @@ def main():
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+            try:
+                ref_prefix, os.environ['GIT_DIR'] = data.split('gitdir: ', 1)
```
Honestly, I don't understand why you'd want to change this env var at all. If the repo already has a separate git dir it will figure it out automatically. The only step git looks at this var is clone/init phase, for which there's an appropriate `--separate-git-dir` CLI argument, which is preferable over env var.

cmrd-senya replied:
Because in order to set `--separate-git-dir` I would have to modify every function call and definition in this file which calls a git command

webknjaz replied:
This place in code **does not affect** cloning/initing, while operating on **existing repo**, which makes it pointless. Also, it would be architecturally wrong to do it from some random phase of module execution. We would still have to bypass it to clone/init anyway.

cmrd-senya replied:
Why? The problem at #30034 is related to existing repo only.

cmrd-senya replied:
The point of this PR is not to support cloning/initing with a separate git dir, the point is to support updating the existing repositories with separate git dir.

webknjaz replied:
1) It does make almost no sense to support this feature of Git only partially.
2) `GIT_DIR` env var does not affect fetch, since Git would already parse `gitdir: ...` from `.git` file to figure out where the repo is located.
3) Also, the very first snippet (and the issue description) in #30034 showcases `clone` happening there. Which happens because of `not os.path.exists(gitconfig)`, not because of env var.

cmrd-senya replied:
> Also, the very first snippet (and the issue description) in #30034 showcases clone happening there. Which happens because of not os.path.exists(gitconfig), not because of env var.

Clone is happening right because the ansible git module fails to detect the existing cloned repository. If you read the issue description carefully it says:

>  if I try to update using the git module it fails with this error

webknjaz replied:
Yes, but it is still unrelated to the env var.

webknjaz reviewed (changes requested):
So to summarize comments:
1) we need to support `separate-git-dir` module argument
2) have docs/examples for that
3) _probably_ drop setting an env var
4) cover all of that with tests including negative ones

webknjaz commented:
You'll need to bypass an arg into `clone` function and also I've noticed some similar config lookup in `get_head_branch` function, so it might also need changes.

Besides above, you'll need to tests this combined with submodules, since I've got a feeling that they might get broken with such changes. It seems like setting `GIT_DIR` var might be already breaking that.

cmrd-senya commented:
> we need to support separate-git-dir module argument

I disagree. My intention is to fix #30034. This PR handles the case where the existing repo with separate git dir is updated, which can be detected automatically by examining `.git` type and contents. I'm currently not interested in implementing support for cloning new repositories with separate git dirs in ansible.

webknjaz commented:
Okay, I can agree with not adding a new module argument in this PR, but I don't agree to accept it without proper testing and with env var modification (which is not needed).

Will be waiting for those then and will do the next round of review afterwards.

cmrd-senya commented:
> but I don't agree to accept it without proper testing and with env var modification (which is not needed).

Does it mean I have to change every function which calls git commands? Will I have to change every such function's argument list and pass separate git dir value down stream everywhere? Or is there a way to store the separate git dir value globally for this module?

webknjaz commented:
I've just looked through `get_head_branch` function and it repeats part of logic for pulling out the gitdir from config, I suggest that it could be moved to a standalone function and reused to avoid further code diversity in future.

webknjaz commented:
> Does it mean I have to change every function which calls git commands? Will I have to change every such function's argument list and pass separate git dir value down stream everywhere? Or is there a way to store the separate git dir value globally for this module?

No, your guess about role of env var was misleading, we just need to drop it, no other modifications needed.

cmrd-senya commented:
> No, your guess about role of env var was misleading, we just need to drop it, no other modifications needed.

Oh, right, it seems to work without it :+1:

webknjaz commented:
Yes, as I told it only affects clone/init, which was very confusing to me and I had to research it myself first.

mattclay commented:
CI failure in integration tests `test/integration/targets/git/tasks/separate-git-dir.yml:15`:

> failure: Current repo does not have a valid reference to a separate Git dir or it refers to the invalid path

Commit:
git: separate-dir: add tests for failures

Test examples for failure cases for git module with git repositories
with separate git directory.

Commit:
git: separate-dir: fix the case handling

Commit:
git: separate-dir: use temporary directory for git dir

cmrd-senya commented:
Updated, please review

webknjaz commented on test/integration/targets/git/tasks/separate-git-dir.yml:
```diff
@@ -0,0 +1,64 @@
+    dest: "{{ checkout_dir }}"
+
+- name: SEPARATE-GIT-DIR | set git dir to non-existent dir
+  shell: "echo gitdir: /dev/null/non-existent-dir > .git"
```
We also need to test the case where `gitdir: ...` is on the other line.

cmrd-senya replied:
Like this https://github.com/ansible/ansible/pull/38016/commits/82430f6c2d0c80475177cbb2dbae0857606abeeb ?

webknjaz replied:
Except it should succeed

webknjaz replied:
That's why I asked you to reuse existing code: https://github.com/ansible/ansible/pull/38016#issuecomment-387800545

cmrd-senya replied:
Okay

webknjaz replied:
To be clear: don't copy-paste blindly, merge the best parts of the approaches ;)

webknjaz commented on test/integration/targets/git/tasks/separate-git-dir.yml:
```diff
@@ -0,0 +1,81 @@
+- name: SEPARATE-GIT-DIR | check update has failed
+  assert:
+    that:
+      - result is failed
```
it should not fails if `.gitdir` contains a list of `key: value`

cmrd-senya replied:
I can't find any reference that proves that. Git documentation [says](https://git-scm.com/docs/gitrepository-layout#_description):
>  Also you can have a plain text file .git at the root of your working tree, containing gitdir: <path> to point at the real directory that has the repository.

And no information about the key-value list.

I also tried a few examples, including
```
key: value
gitdir: ...path
```
and (with newline at the front):
```

gitdir: ...path
```

And all of the examples fail at the repository update stage with the message from git: `fatal: Invalid gitfile format: .git`. Can you please provide any correct examples of key-value lists for the `.git` file which will work with Git?

webknjaz replied:
Not really, but since that other snippet loops through key-value lines there must be something working this way. I'll try investigating this.

webknjaz replied:
It looks like the best way of checking integrity of the repo is [`git-fsck`](https://git-scm.com/docs/git-fsck), but I don't thing that it's needed here now.

webknjaz replied:
But you are right, it does contain a validation for a file immediately starting with `gitdir: `: https://github.com/git/git/blob/master/setup.c#L592

webknjaz replied:
So now we know that it should only contain `gitdir: ...path` and nothing more in that file. Trailing EOLs are being stripped off by git.

cmrd-senya commented:
But don't you think we still need to deduplicate the code with `get_head_branch`?

webknjaz commented:
Right, that's why I haven't merged it yet.

cmrd-senya commented:
I was confused because you approved the PR. I already made the changes locally, but I need time to fix them according to our discoveries about `.git` file. I'll do that tomorrow.

Commit:
git: reuse separate git dir code as get_repo_path

cmrd-senya commented:
Update

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -595,14 +595,7 @@ def is_not_a_branch(git_path, module, dest):
-    associated with.  In the case of a detached HEAD, this will look
-    up the branch in .git/refs/remotes/<remote>/HEAD.
-    '''
+def get_repo_path(module, dest, bare):
```
Returning an error on Ansible module protocol level shouldn't be mixed up with just identifying a directory.
Use exceptions to communicate failures to the calling code and process them there, outside of this function.
So no `module` argument is needed here — let's separate different abstraction layers.

cmrd-senya replied:
BTW, it feels like calling `get_repo_path` for the second time for the same `dest` is somehow redundant. `get_repo_path` is called in `main` anyway. Is there a way to save the result after the first invocation and use it later?

webknjaz replied:
Yes, it's called [`lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache), but it's not available in stdlib of Python 2.7, so we cannot use it (unless we vendor [it's backport](https://pypi.org/p/backports.functools_lru_cache) into the project like `six` distribution).

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -611,23 +604,39 @@ def get_head_branch(git_path, module, dest, remote, bare=False):
     if os.path.isfile(repo_path):
```
Now that we support separate git dir the comment above requires extension.

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -611,23 +604,39 @@ def get_head_branch(git_path, module, dest, remote, bare=False):
     if os.path.isfile(repo_path):
         try:
             git_conf = open(repo_path, 'rb')
```
This is a forgotten, unclosed and unnecessary file descriptor. We need to wipe this line.

webknjaz committed:
Drop forgotten file descriptor

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -611,23 +604,39 @@ def get_head_branch(git_path, module, dest, remote, bare=False):
-        except (IOError, AttributeError):
+            if not os.path.isdir(repo_path):
+                raise ValueError('%s is not a directory' % repo_path)
+        except (IOError, AttributeError, ValueError) as err:
```
Why `IOError` or `AttributeError` can happen? I can see only `ValueError` occuring.

webknjaz committed:
Separate failure processing from logic

webknjaz commented on lib/ansible/modules/source_control/git.py:
```diff
@@ -603,31 +626,16 @@ def get_head_branch(git_path, module, dest, remote, bare=False):
-            return ''
+    try:
+        repo_path = get_repo_path(dest, bare)
+    except (IOError, ValueError) as err:
```
@cmrd-senya `IOError` can happen in case of insufficient privileges or similar issues while accessing fs, but I cannot see `AttributeError` appearing. Do you recall anything causing it?

cmrd-senya replied:
I don't. `AttributeError` was present in the original code and I don't think I ever saw it being thrown.

Commit:
git: update the comment about the separate git dir

cmrd-senya commented:
Updated

cmrd-senya commented:
@webknjaz, thank you for your assistance!

webknjaz commented on test/integration/targets/git/tasks/main.yml:
```diff
@@ -36,3 +36,6 @@
 - include_tasks: archive.yml
+- include_tasks: separate-git-dir.yml
+  when:
+    - git_version.stdout is version("1.7.5", '>=')
```
I missed that it should've been `git_version.stdout | version_compare("1.7.5", '>=')`

cmrd-senya replied:
I copied that from the line 34
